datafusion/expr/src/expr.rs

```diff
@@ -267,6 +267,28 @@ impl BinaryExpr {
     }
 }
 
+impl Display for BinaryExpr {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        // Put parentheses around child binary expressions so that we can see the difference
+        // between `(a OR b) AND c` and `a OR (b AND c)`. A future improvement here would be
+        // to only insert parentheses when needed, based on operator precedence. For example,
+        // `(a AND b) OR c` and `a AND b OR c` are equivalent and the parentheses are not
+        // necessary
+        if matches!(*self.left, Expr::BinaryExpr(_)) {
+            write!(f, "({})", self.left)?;
+        } else {
+            write!(f, "{}", self.left)?;
+        }
+        write!(f, " {} ", self.op)?;
+        if matches!(*self.right, Expr::BinaryExpr(_)) {
+            write!(f, "({})", self.right)?;
+        } else {
+            write!(f, "{}", self.right)?;
+        }
+        Ok(())
+    }
+}
```

Comment:
@Ted-Jiang @alamb  This is the main change. Would be good to get some initial feedback here before I update the remaining tests.

Reply:
I think to avoid too much parentheses we have to take into account the operator precedence.

E.g. not put parentheses in these expressions:

`a AND b OR c AND d`
`a * 10 + a * 3`

etc.

Reply:
I see you already added this as a comment

Reply:
I am torn between doing the bare minimum (the current PR changes) and implementing this fully with operator precedence. This would be much more work, and we would need comprehensive tests. Maybe there is some middle ground where we respect operator precedence for the common cases (AND/OR and +.-.*./) and just wrap anything else in parentheses. We can then improve this over time. I'll try this approach tomorrow.

Reply:
I guess all binary operations are **left-associative** currently, which could also be taken into account to avoid too many parentheses.

Reply:
I am going to update this PR to do the right thing based on the precedence of the operators

Reply:
@Dandandan @HaoYang670 I have now reimplemented this to only insert parentheses when needed. PTAL when you can.